New: Enhance Reference Range (Fixes #267)

SharathAlur opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**What is the purpose of this pull request? (Check any that's applicable)**

-   [x] Feature
-   [ ] Bug fix
-   [ ] Dependency updates
-   [ ] Documentation updates
-   [ ] Build related changes
-   [ ] Changes an existing functionality
-   [ ] Other, please explain:

**Does this introduce a breaking change?**

-   [ ] Yes
-   [x] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

**What changes did you make? (Give an overview)**
Provided a feature for consumer to provide region independently per value in the graph.

**Add any screenshots**
Value region with line type LINEAR (https://engineering.cerner.com/carbon-graphs/#/regions/line/value-level-region)
<img width="1349" alt="Line Value Region" src="https://user-images.githubusercontent.com/22268936/86785873-e34c5980-c080-11ea-8021-c5d4396f8648.png">

Value region with line type SPLINE (https://engineering.cerner.com/carbon-graphs/#/regions/spline/valueRegion)
<img width="1349" alt="Spline Value Region" src="https://user-images.githubusercontent.com/22268936/86785882-e7787700-c080-11ea-9d86-590e94a8d31d.png">

Value region in Paired result (https://engineering.cerner.com/carbon-graphs/#/regions/lpaired-result/value-level-region)
<img width="1349" alt="Paired Result Region" src="https://user-images.githubusercontent.com/22268936/86785886-e8a9a400-c080-11ea-9f80-64221f3517ef.png">

Multiple dataset in a graph (https://engineering.cerner.com/carbon-graphs/#/regions/line/multiple-line)
![Multi_DataSet.gif](https://user-images.githubusercontent.com/22268936/87901117-d150ba80-ca73-11ea-9237-89dfaf54b59a.gif)

Veera81 commented:
@SharathAlur Thanks for adding a new feature to cabon-graphs. Is there any issue logged for this in [carbon-issues](https://github.com/cerner/carbon-graphs/issues)

SharathAlur commented:
No @Veera81. There is no issue logged in the gitHub. I worked based on [Jira](https://jira3.cerner.com/browse/CARBON-113)

Veera81 commented:
@SharathAlur Can you log an issue for just visibility so that consumer will know that we are working on this enhancement

SharathAlur commented:
@Veera81 I have created an issue and updated the PR title.

Veera81 commented:
@SharathAlur Thanks for creating the issue

AshishMotanamGurunadham commented on src/main/js/controls/Line/Line.js:
```diff
@@ -158,23 +169,33 @@ class Line extends GraphContent {
-        if (utils.notEmpty(this.dataTarget.regions)) {
+        if (
+            utils.notEmpty(this.dataTarget.regions) ||
+            !utils.isEmptyArray(this.dataTarget.valueRegionSubset)
```
Since we have notEmpty method in utils. Can we use it in here instead of !isEmptyArray

SharathAlur replied:
valueRegionSubset is an array. utils.notEmpty checks only object or string. So I have used utils.isEmptyArray.

AshishMotanamGurunadham commented on src/main/js/controls/Line/Line.js:
```diff
@@ -158,23 +169,33 @@ class Line extends GraphContent {
-            hideAllRegions(graph.svg);
             graph.config.shouldHideAllRegion = true;
         }
+        if (graph.config.shouldHideAllRegion) {
+            hideAllRegions(graph.svg);
+        }
```
May i know what is the purpose of using separate if statement? Can we just call hideAllRegions method directly in else ?

SharathAlur replied:
While developing I found an issue.
If there are more than two lines and all has regions in it. Ideally the graph should hide all regions. Instead here it was hiding only first two regions. And other regions were being displayed.
So I gotta move it out of that if condition to fix this issue.

AshishMotanamGurunadham commented on src/main/js/controls/Line/helpers/helpers.js:
```diff
@@ -250,6 +250,77 @@ const draw = (scale, config, canvasSVG, dataTarget) => {
+            !utils.isEmpty(value.region.end)
+        ) {
+            // If the color is different, then move it to new region set
+            if (previousColor !== value.region.color) {
```
where will we get the value for previous color from? looks like it just declared in line 269

SharathAlur replied:
previous color is just a temp storage of previously evaluated region. This condition I have added to check if there is any change in the color between the value regions.

AshishMotanamGurunadham commented:
can you add an image showing how regions look like if we have more than one data set present in graph.

AshishMotanamGurunadham commented on dev/examples/controls/line.js:
```diff
@@ -31,7 +31,7 @@ const regions = [
     const lineDefault = Carbon.api.graph(getDemoData(`#${id}`, "LINE_DEFAULT"));
     lineDefault.loadContent(
-        Carbon.api.line(getDemoData(`#${id}`, "LINE_DEFAULT").data[0])
+        Carbon.api.line(getDemoData(`#${id}`, "LINE_DEFAULT").data[2])
```
May i know the reason why did you changed the content in here. i guess it is changing the shape of our basic line graph.

![image](https://user-images.githubusercontent.com/63670637/87202598-7c1fe680-c2c6-11ea-8f97-1ab268783b59.png)

SharathAlur replied:
I had changed for some testing purpose. I reverted now.
https://github.com/SharathAlur/carbon-graphs/commit/788fde01aa11c505624e86e5a4ce9bc7da6d0813#diff-8ef55c17e956a86bd177242c976fc979L34

AshishMotanamGurunadham commented on src/main/js/controls/Line/helpers/helpers.js:
```diff
@@ -250,6 +250,77 @@ const draw = (scale, config, canvasSVG, dataTarget) => {
+        ) {
+            // If the color is different, then move it to new region set
+            if (previousColor !== value.region.color) {
+                if (valueRegion.values.length > 0) {
```
Can you add comment for all if statements which can explain the functionality?

AshishMotanamGurunadham commented on src/main/js/controls/Line/helpers/helpers.js:
```diff
@@ -250,6 +250,77 @@ const draw = (scale, config, canvasSVG, dataTarget) => {
+    });
+    if (valueRegion.values.length > 0) {
+        valueRegionSubset.push(valueRegion);
+    }
```
instead of writing extra if statement can we add values directly in valueRegionSubset ?

SharathAlur replied:
I do not want to push the region without any values. So, I am checking before adding to valueRegionSubset. Now I have moved two if condition to one place.

AshishMotanamGurunadham commented on src/main/js/controls/Line/helpers/helpers.js:
```diff
@@ -250,6 +250,77 @@ const draw = (scale, config, canvasSVG, dataTarget) => {
+    // This is to cover the start and end data value with region.
+    valueRegionSubset.forEach((region) => {
+        valueRegions.push(region);
+        if (region.values.length > 1) {
```
can we have multiple values in one region? if not why will we have length more than 1?

AshishMotanamGurunadham commented on src/main/js/controls/PairedResult/PairedResult.js:
```diff
@@ -222,7 +232,9 @@ class PairedResult extends GraphContent {
-                )
+                ),
+                this.dataTarget.yAxis,
+                !!utils.notEmpty(this.dataTarget.valueRegionSubset)
```
nit: can you remove !! in here.

SharathAlur replied:
Updated: https://github.com/SharathAlur/carbon-graphs/commit/788fde01aa11c505624e86e5a4ce9bc7da6d0813

SharathAlur commented:
@Veera81 @sdadn @mraplumb @AshishMotanamGurunadham I have addressed all the review comments. Can you please review?
Also please add appropriate labels/tags for UX review.

Veera81 commented:
@SharathAlur Looks like in the PR description, first and the second screenshot is the same. The second screenshot supposed to be a different one?

Can you add a comment what dev examples should the UX review should be looking at?

SharathAlur commented:
@Veera81 The data for the first and second screenshot is same, but the line graph type is different. First and second screenshots are Linear and Spline type respectively. I have added Spline to showcase the value region inline with the line interpolation type.

I have added following examples in the dev-site:
http://localhost:9991/#/regions/line/value-level-region
http://localhost:9991/#/regions/paired-result/value-level-region
http://localhost:9991/#/regions/spline/valueRegion

AshishMotanamGurunadham commented:
> can you add an image showing how regions look like if we have more than one data set present in graph.

Can you please add an image for the changes you made with reference range provided for multiple datasets?

mraplumb commented:
@SharathAlur can you update the PR description to specifically note which examples in the dev-site where these updates can be seen when your code changes are built?

mraplumb commented:
@SharathAlur I see you responded to the comment indicating the examples in the dev site where this will be able to be seen.  Can you update your PR description with this information so they are easily found by others?

mraplumb commented on src/main/js/controls/Line/helpers/helpers.js:
```diff
@@ -250,6 +250,79 @@ const draw = (scale, config, canvasSVG, dataTarget) => {
+ * Extracts regions inside value and prepare a list of value region
+ *
+ * @param {object} dataTarget - Data points object
+ * @param {Function} getXDataValues - checks if the value is valid and etract value for x-axis
```
nit: etract -> extract?

SharathAlur replied:
Updated

mraplumb commented on src/main/js/controls/Line/helpers/helpers.js:
```diff
@@ -250,6 +250,79 @@ const draw = (scale, config, canvasSVG, dataTarget) => {
+                if (valueRegion.values.length > 0) {
+                    valueRegionSubset.push(valueRegion);
+                }
+                previousColor = value.region.color;
```
is this line necessary?  It should be taken care of below on line 287

SharathAlur replied:
I agree. It is not required. I have removed this line.

SharathAlur commented:
@mraplumb I  have updated the PR with the link to the demo as, where exactly the examples are present. The links will proper example once the PR is merged.
@AshishMotanamGurunadham When there is multiple data set, carbon-graphs will hide all the regions. It will be highlighted on hover of particular legend.

SharathAlur commented on src/test/unit/controls/Line/LineRegion-spec.js:
```diff
@@ -84,6 +84,97 @@ describe("Line - Region", () => {
+                    color: "#f4f4f4"
+                };
+            });
+            it("When the values contain region object", () => {
```
I was reading the statements from describe. Now I have updated the tests.
https://github.com/cerner/carbon-graphs/pull/258/commits/fc9e8c9335234c69dc6d9913b7a9fcc71e36d1b8#diff-5e053a5484d9ccc20008ced47434119bR97

SharathAlur commented on docs/controls/Line.md:
```diff
@@ -222,10 +222,13 @@ Each line can have a legendOptions object in [Values](#values) level.
-Each data-set can have 1 or more regions. `start` and `end` is necessary for rendering a horizontal area.
+Each data-set can have 1 or more regions.
+Each value can have a region object.
+`start` and `end` is necessary for rendering a horizontal area.
```
I have updated the doc for line and paired result. As only these two graph types supports value region.
https://github.com/cerner/carbon-graphs/pull/258/commits/fc9e8c9335234c69dc6d9913b7a9fcc71e36d1b8#diff-ca5f8f98fbfac861a5087764810586d5R232

SharathAlur commented:
@jeremyfuksa I have addressed all the review comments and got an approval. Can you please start the UX review? Let me know if there are any concerns.

SharathAlur commented:
@jeremyfuksa Any update on the UX review?

SharathAlur commented on src/test/unit/controls/PairedResult/PairedResultRegion-spec.js:
```diff
@@ -174,6 +174,95 @@ describe("Paired Result - Region", () => {
+                    end: 70
+                };
+            });
+            it("If value contains region object", () => {
```
I have updated the descriptions.

jeremyfuksa commented:
In the value region line example, the darker gray against the blue data set does not meet the 3:1 color contrast ration required. The dark gray(`#A6AAAB`) must be at least `#B8BBBC` in order to pass.

SharathAlur commented:
@jeremyfuksa To fix the contrast I have changed the line colour to black. With that all the colour contrast ratio is more than 3.

ryanthemanuel commented:
:tada: This PR is included in version 2.11.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@cerner/carbon-graphs/v/2.11.0)
- [GitHub release](https://github.com/cerner/carbon-graphs/releases/tag/v2.11.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
